agent/api/ecsclient/client.go

```diff
@@ -155,16 +155,43 @@ func (client *APIECSClient) registerContainerInstance(clusterRef string, contain
 	// Add additional attributes such as the os type
 	registrationAttributes = append(registrationAttributes, client.getAdditionalAttributes()...)
 	registerRequest.Attributes = registrationAttributes
+	registerRequest = client.setInstanceIdentity(registerRequest)
+
+	resources, err := client.getResources()
+	if err != nil {
+		return "", err
+	}
+
+	registerRequest.TotalResources = resources
+	resp, err := client.standardClient.RegisterContainerInstance(&registerRequest)
+	if err != nil {
+		seelog.Errorf("Unable to register as a container instance with ECS: %v", err)
+		return "", err
+	}
+	seelog.Info("Registered container instance with cluster!")
+	err = validateRegisteredAttributes(registerRequest.Attributes, resp.ContainerInstance.Attributes)
+	return aws.StringValue(resp.ContainerInstance.ContainerInstanceArn), err
+}
+
+func (client *APIECSClient) setInstanceIdentity(registerRequest ecs.RegisterContainerInstanceInput) ecs.RegisterContainerInstanceInput {
+	instanceIdentityDoc := ""
+	instanceIdentitySignature := ""
+
+	if client.config.NoIID {
+		seelog.Info("Fetching Instance ID Document has been disabled")
+		registerRequest.InstanceIdentityDocument = &instanceIdentityDoc
```

Comment:
Why do we assign empty string values? can't these fields be nil?

Reply:
the client we use expects string pointers

Reply:
I tried to use this solution using the following configuration file:
```json
{
 "NoIID": true
}
```
I use following command to run:
```
docker run --name ecs-agent --rm --privileged -v /var/run/docker.sock:/var/run/docker.sock -p 127.0.0.1:51678:51678 -e ECS_LOGFILE=/dev/stdout -e ECS_LOGLEVEL=info -e AWS_ACCESS_KEY_ID=... -e AWS_SECRET_ACCESS_KEY=... -e AWS_DEFAULT_REGION=eu-west-2 -e ECS_CLUSTER=ecs-cluster -e "ECS_AGENT_CONFIG_FILE_PATH=/etc/ecs/ecs.config.json" -v /tmp/config.json:/etc/ecs/ecs.config.json amazon/amazon-ecs-agent:latest -blackhole-ec2-metadata
```
I receive following logs:
```
2019-11-03T14:02:30Z [INFO] Loading configuration_CONFIG_FILE_PATH=/etc/ecs/ecs.config.json" -v /tmp/config.json:/etc2019-11-03T14:02:30Z [INFO] Image excluded from cleanup: amazon/amazon-ecs-agent:latest
2019-11-03T14:02:30Z [INFO] Image excluded from cleanup: amazon/amazon-ecs-pause:0.1.0
2019-11-03T14:02:30Z [WARN] Unable to fetch user data: blackholed
2019-11-03T14:02:30Z [INFO] Amazon ECS agent Version: 1.32.1, Commit: 4285f58f
2019-11-03T14:02:30Z [WARN] Unable to access EC2 Metadata service to determine EC2 ID: blackholed
2019-11-03T14:02:30Z [INFO] Creating root ecs cgroup: /ecs
2019-11-03T14:02:30Z [INFO] Creating cgroup /ecs
2019-11-03T14:02:30Z [INFO] Checkpointing not enabled; a new container instance will be created each time the agent is run
2019-11-03T14:02:30Z [INFO] Event stream ContainerChange start listening...
2019-11-03T14:02:30Z [WARN] Unable to obtain Outpost ARN from EC2 Metadata: blackholed
2019-11-03T14:02:30Z [INFO] Registering Instance with ECS
2019-11-03T14:02:30Z [INFO] Fetching Instance ID Document has been disabled
2019-11-03T14:02:30Z [INFO] Remaining mem: 15937
2019-11-03T14:02:30Z [ERROR] Unable to register as a container instance with ECS: InvalidParameterException: The identity document and identity document signature were not valid.
	status code: 400, request id: c0c3ae52-c5fc-43ef-9da5-3679636f8cb0
2019-11-03T14:02:30Z [ERROR] Error registering: InvalidParameterException: The identity document and identity document signature were not valid.
	status code: 400, request id: c0c3ae52-c5fc-43ef-9da5-3679636f8cb0
2019-11-03T14:02:30Z [CRITICAL] Instance registration attempt with an invalid parameter
```

According to the documentation https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_RegisterContainerInstance.html the fields "instanceIdentityDocument" and "instanceIdentityDocumentSignature" are not required. They don't seem to accept empty values.

The above suggests that the empty text value is sent and then verified, which causes the above error.